var/spack/repos/builtin/packages/python/package.py

```diff
@@ -202,6 +204,20 @@ def patch(self):
             r'\1setup.py\2 --no-user-cfg \3\6'
         )
 
+    @when('@:3.6.999+ssl')
+    @run_before('build')
```

Comment:
Does @when and @run_before work in conjunction with each other? I feel like I've had problems with this in the past, you should make sure it doesn't always run or never run.

Reply:
I retried the install with variants that should both pass and fail these `@when` conditions (specifically, `python@2.7.12+ssl` and `python@2.7.12~ssl`) and it looks like `@run_before` and `@when` don't play well together. In particular, it looks like the `@when` gets ignored and the function is always run, which explains why I didn't encounter problems before when I was running purely `+ssl` variants.

This being the case, what's the best way to run both of these patches? I don't think multiple `patch` functions will work unless they span disjoint specs (see `qt`), so is the only option to write a general `patch` function that has conditional clauses for these two different cases?

Reply:
Yeah, I would recommend a single patch function with two if statements, one for each version range.

Reply:
See https://github.com/spack/spack/issues/12736. If we encounter a lot of cases like this we might raise the impact level of the bug.